tests/e2e/mock_llm.py

```
@@ -331,73 +337,208 @@ def _next_llm_trace_response(
+        ]
     if not isinstance(value, dict):
         return value
+    if "$trace_result" not in value:
+        return {
+            key: _resolve_trace_result_bindings(item, tool_results)
+            for key, item in value.items()
+        }
+    if set(value) != {"$trace_result"}:
+        raise ValueError(
+            "$trace_result marker object must contain only $trace_result"
+        )
```

Ruff (0.15.21)

[warning] Avoid specifying long messages outside the exception class

(TRY003)